core/src/main/java/com/linecorp/armeria/common/logging/ContentPreviewerFactory.java

```diff
@@ -0,0 +1,151 @@
+    }
+
+    ContentPreviewerFactory DISABLED = (ctx, headers) -> ContentPreviewer.DISABLED;
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(MediaType contentType, Supplier<ContentPreviewer> supplier) {
+        return of((ctx, headers) -> {
+            final MediaType type = headers.contentType();
+            return type != null && type.is(contentType);
+        }, supplier);
+    }
+
+    static ContentPreviewerFactory of(String contentType, Supplier<ContentPreviewer> supplier) {
+        return of(MediaType.parse(contentType), supplier);
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(BiPredicate<RequestContext, HttpHeaders> predicate,
+                                      Supplier<ContentPreviewer> supplier) {
+        return (ctx, headers) -> {
+            if (predicate.test(ctx, headers)) {
+                return supplier.get();
+            }
+            return ContentPreviewer.DISABLED;
+        };
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(ContentPreviewerFactory factory1, ContentPreviewerFactory factory2) {
+        if (factory1 == DISABLED) {
+            return factory2;
+        }
+        if (factory2 == DISABLED) {
+            return factory1;
+        }
+        if (factory1 instanceof ContentPreviewerFactoryCombinable) {
+            return ((ContentPreviewerFactoryCombinable) factory1).combine(factory2);
+        }
+        if (factory2 instanceof ContentPreviewerFactoryCombinable) {
+            return ((ContentPreviewerFactoryCombinable) factory2).combine(factory1);
+        }
+        return new CompositeContentPreviewerFactory(Arrays.asList(factory1, factory2));
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(ContentPreviewerFactory... factories) {
+        final List<ContentPreviewerFactory> factoryList = new ArrayList<>();
+        ContentPreviewerFactory combined = DISABLED;
+        for (ContentPreviewerFactory factory : factories) {
+            if (factory == DISABLED) {
+                continue;
+            }
+            if (factory instanceof ContentPreviewerFactoryCombinable) {
+                combined = ((ContentPreviewerFactoryCombinable) factory).combine(combined);
+            } else if (combined instanceof ContentPreviewerFactoryCombinable) {
+                combined = ((ContentPreviewerFactoryCombinable) combined).combine(factory);
+            }
+            factoryList.add(factory);
+        }
+        if (combined != DISABLED) {
+            factoryList.add(combined);
+        }
+        if (factoryList.size() < 2) {
+            return factoryList.isEmpty() ? DISABLED : factoryList.get(0);
+        }
+        return new CompositeContentPreviewerFactory(factoryList);
+    }
+
+    /**
+     * TODO: AddJavadocs.
+     */
+    static ContentPreviewerFactory of(Map<MediaType, Supplier<ContentPreviewer>> map) {
```

Comment:
`Map<MediaType, ? extends Supplier<? extends ContentPreviewer>>`?

Reply:
I'm not sure if I can do that because `? extends Supplier<? extends ContentPreviewer>` cannot be casted as `Supplier<ContentPreviewer>`

Reply:
Had a chat in offline to show it is possible indeed.